### ISSUE TITLE
Comparison with Webix

### ISSUE THREAD

**Issue by dandv** [vitmalina/w2ui#888]
So many UI libraries out there! As one of the testimonials says,

> I was searching for a slick and simple, but also user-friendly UI library. I tried Polymer, Angular, OpenUI5, jqwidget, - but they do not satisfy my needs. I don't want to declare the whole UI in html (templates) but instead will create ALL UI elements programmatically (in an easy way). With w2ui this can be done really easy and the UI elements looks really nice.   
> -- Gerald Leeb

That user hasn't tried Webix though.

I'm sure you've looked at Webix and DHTMLX before starting W2UI. How would you say W2UI compares with [Webix](http://webix.com), aside from the smaller number of widgets, but freer license (MIT)?

**Comment by vitmalina** [comment]
I think I am wrong guy to answer this questions. I am biased towards w2ui as I have been working on it since 2007-ish. I see you are in bay area just like me. I can come talk to your team and do a presentation about w2ui.

**Comment by hansgru** [comment]
> How would you say W2UI compares with Webix, aside from the smaller number of widgets, but freer license (MIT)?

As a user of all above frameworks and also a few others (e.g. ExtJS, KendoUI,qooxdoo, etc), I would say that except the missing IE8 support, W2UI is the best in the following categories:
- **learnability** - with interactive jsfiddle examples (now others have this too), w2ui is still better (even if other have even books or a bigger community). JS developers can really get productive with w2ui after a day already.
- **grid performance** - so far the best for handling huge amount of data. ExtJS can do more, but it has a much worse learning curve, and the bugs plague components for years.
- **speed of fixing bugs** - so far the best experience with w2ui (other frameworks do incompatible rewrites instead of fixing bugs, others will try to convince you it's a feature) 
- **prototyping** - even if there's no GUI builder (like other frameworks have), because of the many examples in fiddle form, it's even faster to prototype stuff than with GUI builders.

We haven't found Webix to be better suited than the concurrence at any of the important tasks we usually require.

Areas where w2ui does not shine (yet ?) :
 - Theming
 - accessibility ( KendoUI is the best here: http://docs.telerik.com/kendo-ui/accessibility/accessibility-overview )
 - a few more "business" components (see ExtJS )

Of course, all this not including the price :) (for some this is quite high, the support is miserable, and after a few years of wasted money we had to cancel most of our subscriptions).

Some directions we would like w2ui to consider:
 - React for rendering
 - ClojureScript integration

**Comment by vitmalina** [comment]
@hansgru - thanks for the feedback. Appreciate it. I agree with your points where w2ui does not shine. Theming is something that could possibly be improved. It has not been too much of pain point for users as I have not heard much feedback about it. There is a dark theme, but I admit, it has not been updated for a while. Since all css are broken into components and are in css preprocessor, it should provide a good start to implement custom themes. Most of the color and style options are variables in LESS. It would be nice to have a theme roller, if someone would take time to do it.

As for accessibility, you are right too. I have mainly ignored it. I have attended Fluent a month ago and there was a session on accessibility. I admit I was impressed by the speaker as she did excellent job and only 10 min into her talk I have realized that she was blind. She did really well. It might serve as a catalyst to work on accessibility.

As for React, I have communicated with a developer from React team and he did a test where he took the 1 million records example from w2ui and reimplemented it with React. He said that he almost got the same speed of rendering with React, which was impressive. But I noticed the speed of rendering for w2ui was a bit faster. The speed is important to me and I am constantly looking to improve it. 

As for ClosureScript, I am not sure I understand what I can do to make it smoother. Recently, I have talked to some developers that used meteor and they wanted tighter integration with meteor, but I am not sure too, what kind if integration should it be. W2UI is a set of UI widgets, it is not a framework, and it is indifferent to other libraries you use with it. It should play nicely with anything out there. 

As for direction, here is main points I am working on in last few months

1. Feature developments - more options for controls, more options in the grid (frozen columns, nested records, colspan, excel like grid, etc). New features in popup, toolbar, sidebar. I will consider business features that you say extjs has.

2. Smother user experience. For example, I have been working in past months to make keyboard navigation better. Now grid has focus/blur events and it has a tab stop. It can get or loose focus. It has been mainly working but still has issues with sugrids. Also, popup has been improved with better keyboard navigation support, as well as Sidebar. When popup is open, it will focus first button and clicking tab key with cycle between controls inside popup and will not break out.

3. I am considering improving w2form and implement 2-way data binding (not in close future). I have heard many comments where people liked that forms could auto generate default template. I think it is important to have faster development cycle. I have been working on a project where I used rivetsjs and I really like how easy it is to use it and it never gets in my way.

4. I am keeping my eyes and ears open for new ECMA6 standards, web components, and other trends in modern web developments as I think it is importable to keep with the trends.
 
Most of the new features that w2ui gets are due to business needs of the projects I am working on. It has been a blessing for me to be able to work on many projects where w2ui was a central building block and learn new ways where it could bring value and implement new features. I have heard from so many people and have seen dozens of projects done with w2ui all across the globe and it has been revelational to see how people use it.

**Comment by vitmalina** [comment]
I will close this issue, as it is not a bug, but if someone wants to continue discussion - please do.

**Comment by hansgru** [comment]
@vitmalina thank you for your detailed explanation. IMO the "direction" section could make a very good blog post, as a "closed ticket" is mostly ignored :).


> It would be nice to have a theme roller, if someone would take time to do it.

IMO this is not really required. We always tried those from the other frameworks, but really never used them for production, as the "selected" properties are not reversible: i.e. I can't save them to the VCS and later re-import them in the Roller to do changes again, so we always had to start from zero.

A much more practical approach for theming would be:
 - more base/common properties so that there's mostly a single file that needs to be changed
 - better docs
 - better Twitter Bootstrap integration, since many use this instead.


> As for accessibility ...

This is a nice solution for Twitter Bootstrap: https://github.com/paypal/bootstrap-accessibility-plugin and so far we had a very good experience with it.


> As for React, .... But I noticed the speed of rendering for w2ui was a bit faster. The speed is important to me and I am constantly looking to improve it. 

Speed is of course very important, but so far in w2ui only the Grid component is faster. React could be used however for all the other components. At least in our case, it greatly reduced the code we had to write and maintain.


> As for ClosureScript, I am not sure I understand what I can do to make it smoother.

Well, the combination of React and ClojureScript seems to be the magic sauce :).
As I understood, basically the entire model needs to be a tree, than something like:
http://www.infoq.com/news/2014/01/om-react
https://github.com/omcljs/om or 

https://github.com/tonsky/rum or

https://reagent-project.github.io/

can be used, with quite a few advantages. E.g. unlimited undo/redo out of the box being just one of them.

http://blog.scalac.io/2015/04/02/clojurescript-reactjs-reagent.html

> Recently, I have talked to some developers that used meteor and they wanted tighter integration with meteor, but I am not sure too, what kind if integration should it be.

We use Meteor too and it's fantastic, but in many cases it's not really an option, as JS on the server is allowed for building, but in many cases not really accepted in production purposes.

> 2. Smother user experience.  ....

Also when the popup is modal, the background elements should not be focusable or it should not be addressable with the keyboard or navigation.

> 3. I am considering improving w2form and implement 2-way data binding ...

I'm not sure if this would be helpful as it would make everything much more complicated. IMO persistent data structures from ClojureScript could bring much more to the table. They can be used even without ClojureScript: http://swannodette.github.io/mori/

> I have heard from so many people and have seen dozens of projects done with w2ui all across the globe and it has been revelational to see how people use it.

Maybe a section "Who is using w2ui" with "screeshots" and a short description from the users would be even more helpful, and could convince other users easier too?

**Comment by hansgru** [comment]
Also regarding theming, what seems to make it more complicated is that there still seems to be quite a few inline styles.

**Comment by vitmalina** [comment]
yes, it is a goal to completely move inline styles into css.
